File: rust/crates/corlinman-shadow-tester/src/eval.rs
//! Eval-case spec + YAML loader for ShadowTester.
//!
//! Cases live as YAML files under `<eval_set_dir>/<kind>/*.yaml`. The
//! per-kind subdir is the contract: it lets [`load_eval_set`] default
//! `kind` from the path so authors don't have to repeat themselves, and
//! makes `ls memory_op/` the way an operator audits coverage.
//!
//! Step 2 of W1-A: types + loader + fixtures only. Step 3 wires these
//! into the simulator.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use corlinman_evolution::{EvolutionKind, EvolutionRisk};
use serde::{Deserialize, Serialize};

/// One YAML-defined test case for one kind.
///
/// `kb_seed` runs raw SQL against a tempdir copy of `kb.sqlite` before
/// the proposal is shadowed. Normally `INSERT INTO chunks ...`; cases
/// that want a from-scratch fixture can include `CREATE TABLE` first.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvalCase {
    /// Defaults to the YAML file stem in [`load_eval_set`].
    #[serde(default)]
    pub name: String,
    /// Defaults to the directory's kind in [`load_eval_set`].
    #[serde(default)]
    pub kind: Option<EvolutionKind>,
    pub description: String,
    #[serde(default)]
    pub kb_seed: Vec<String>,
    /// `<basename> -> file body` map written into a runner-managed
    /// per-case `<tempdir>/skills/` directory before the simulator runs.
    /// `BTreeMap` keeps YAML serialization stable. Empty for kinds that
    /// don't touch `skills/` (memory_op, tag_rebalance).
    #[serde(default)]
    pub skill_seed: BTreeMap<String, String>,
    pub proposal: ProposalSpec,
    pub expected: ExpectedOutcome,
}

/// Minimum proposal data the simulator needs. Step 3's runner assembles
/// a full `EvolutionProposal` by attaching a generated id + timestamps.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProposalSpec {
    pub target: String,
    pub reasoning: String,
    /// Shadow only fires for medium/high; default High keeps fixtures
    /// in-scope by default.
    #[serde(default = "default_risk")]
    pub risk: EvolutionRisk,
    #[serde(default)]
    pub signal_ids: Vec<i64>,
    /// Unified-diff payload — `skill_update` ships an `__APPEND__`
    /// hunk here. memory_op / tag_rebalance leave it empty (target +
    /// kb_seed are sufficient for those).
    #[serde(default)]
    pub diff: String,
}

fn default_risk() -> EvolutionRisk {
    EvolutionRisk::High
}

/// What the simulator should observe after replaying the proposal.
///
/// `outcome` is the discriminator. New variants extend the enum without
/// breaking existing fixtures because serde tags by field, not order.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "outcome", rename_all = "snake_case")]
pub enum ExpectedOutcome {
    /// memory_op: a merge that consumed `rows_merged` chunks and kept
    /// `surviving_chunk_id` as the canonical row.
    Merged {
        rows_merged: u32,
        surviving_chunk_id: i64,
        #[serde(default = "default_latency_ms_max")]
        latency_ms_max: u64,
    },
    /// memory_op: simulator detected a bogus / unsafe target.
    NoOp {
        #[serde(default = "default_latency_ms_max")]
        latency_ms_max: u64,
    },
    /// tag_rebalance: `merge_tag` executed — the source tag node is
    /// gone and `parent_id` now owns its `chunk_tags` rows.
    TagMerged {
        src_path: String,
        parent_id: i64,
        moved_chunk_count: u32,
        #[serde(default = "default_latency_ms_max")]
        latency_ms_max: u64,
    },
    /// tag_rebalance: target path didn't resolve to a real `tag_nodes`
    /// row — no rows changed.
    TagNoOp {
        #[serde(default = "default_latency_ms_max")]
        latency_ms_max: u64,
    },
    /// skill_update: file appended; final body must contain
    /// `content_includes` as a substring.
    SkillUpdated {
        file: String,
        content_includes: String,
        #[serde(default = "default_latency_ms_max")]
        latency_ms_max: u64,
    },
    /// skill_update: rejected (unsupported diff shape, missing file,
    /// or invalid target).
    SkillNoOp {
        #[serde(default = "default_latency_ms_max")]
        latency_ms_max: u64,
    },
}

fn default_latency_ms_max() -> u64 {
    500
}

/// All cases loaded from one `<dir>/<kind>/` subdir.
#[derive(Debug, Clone)]
pub struct EvalSet {
    pub kind: EvolutionKind,
    pub cases: Vec<EvalCase>,
    pub loaded_from: PathBuf,
}

/// Step-3 fills in the `metrics` shape per simulator. Kept free-form
/// (`serde_json::Map`) so memory_op vs skill_update don't need a shared
/// schema.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvalRunResult {
    pub case_name: String,
    pub passed: bool,
    pub metrics: serde_json::Map<String, serde_json::Value>,
    pub error: Option<String>,
}

/// Loader errors. Empty-set is intentionally an error: a misconfigured
/// path that silently shadows zero cases would look "green" forever.
#[derive(Debug, thiserror::Error)]
pub enum EvalLoadError {
    #[error("eval-set dir missing: {0}")]
    MissingDir(PathBuf),
    #[error("io error reading {path}: {source}")]
    Io {
        path: PathBuf,
        #[source]
        source: std::io::Error,
    },
    #[error("parse failure in {file}: {reason}")]
    ParseFailure { file: PathBuf, reason: String },
    #[error("kind mismatch in {file}: expected {expected:?}, found '{found}'")]
    KindMismatch {
        file: PathBuf,
        expected: EvolutionKind,
        found: String,
    },
    #[error("no eval cases found for kind {kind:?} under {dir}")]
    EmptySet { dir: PathBuf, kind: EvolutionKind },
}

/// Load every `*.yaml`/`*.yml` file from `<eval_set_dir>/<kind>/`.
///
/// Files prefixed with `_` are skipped (drafts). Non-recursive. Cases
/// are sorted by `name` for deterministic ordering across runs.
pub async fn load_eval_set(
    eval_set_dir: &Path,
    kind: EvolutionKind,
) -> Result<EvalSet, EvalLoadError> {
    let dir = eval_set_dir.join(kind.as_str());
    if !tokio::fs::try_exists(&dir)
        .await
        .map_err(|e| EvalLoadError::Io {
            path: dir.clone(),
            source: e,
        })?
    {
        return Err(EvalLoadError::MissingDir(dir));
    }

    let mut entries = tokio::fs::read_dir(&dir)
        .await
        .map_err(|e| EvalLoadError::Io {
            path: dir.clone(),
            source: e,
        })?;
    let mut yaml_files: Vec<PathBuf> = Vec::new();
    while let Some(entry) = entries.next_entry().await.map_err(|e| EvalLoadError::Io {
        path: dir.clone(),
        source: e,
    })? {
        let path = entry.path();
        let Some(file_name) = path.file_name().and_then(|n| n.to_str()) else {
            continue;
        };
        if file_name.starts_with('_') {
            continue;
        }
        let is_yaml = matches!(
            path.extension().and_then(|e| e.to_str()),
            Some("yaml") | Some("yml")
        );
        if !is_yaml {
            continue;
        }
        if !entry
            .file_type()
            .await
            .map_err(|e| EvalLoadError::Io {
                path: path.clone(),
                source: e,
            })?
            .is_file()
        {
            continue;
        }
        yaml_files.push(path);
    }

    let mut cases = Vec::with_capacity(yaml_files.len());
    for file in yaml_files {
        let text = tokio::fs::read_to_string(&file)
            .await
            .map_err(|e| EvalLoadError::Io {
                path: file.clone(),
                source: e,
            })?;
        let mut case: EvalCase =
            serde_yaml::from_str(&text).map_err(|e| EvalLoadError::ParseFailure {
                file: file.clone(),
                reason: e.to_string(),
            })?;

        // Reject explicit-but-wrong kinds; default the unset case to the dir's kind.
        match case.kind {
            Some(found) if found != kind => {
                return Err(EvalLoadError::KindMismatch {
                    file: file.clone(),
                    expected: kind,
                    found: found.as_str().to_string(),
                });
            }
            _ => {
                case.kind = Some(kind);
            }
        }

        if case.name.is_empty() {
            case.name = file
                .file_stem()
                .and_then(|s| s.to_str())
                .unwrap_or("unnamed")
                .to_string();
        }
        cases.push(case);
    }

    if cases.is_empty() {
        return Err(EvalLoadError::EmptySet { dir, kind });
    }

    cases.sort_by(|a, b| a.name.cmp(&b.name));
    Ok(EvalSet {
        kind,
        cases,
        loaded_from: dir,
    })
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn fixtures_root() -> PathBuf {
        // crate-relative: <crate>/tests/fixtures/eval
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("tests")
            .join("fixtures")
            .join("eval")
    }

    async fn write(dir: &Path, name: &str, body: &str) {
        tokio::fs::write(dir.join(name), body).await.unwrap();
    }

    #[tokio::test]
    async fn load_eval_set_returns_missing_dir_when_path_absent() {
        let tmp = TempDir::new().unwrap();
        let err = load_eval_set(tmp.path(), EvolutionKind::MemoryOp)
            .await
            .unwrap_err();
        assert!(matches!(err, EvalLoadError::MissingDir(_)), "got {err:?}");
    }

    #[tokio::test]
    async fn load_eval_set_parses_real_fixtures() {
        let set = load_eval_set(&fixtures_root(), EvolutionKind::MemoryOp)
            .await
            .expect("fixtures should load");
        assert_eq!(set.kind, EvolutionKind::MemoryOp);
        assert_eq!(set.cases.len(), 4, "expected 4 memory_op cases");

        let names: Vec<&str> = set.cases.iter().map(|c| c.name.as_str()).collect();
        assert_eq!(
            names,
            vec![
                "case-001-near-duplicate-merge",
                "case-002-distinct-no-op",
                "case-003-identical-content",
                "case-004-three-way-cluster",
            ]
        );
        assert!(matches!(
            set.cases[0].expected,
            ExpectedOutcome::Merged { .. }
        ));
    }

    #[tokio::test]
    async fn load_eval_set_rejects_malformed_yaml() {
        let tmp = TempDir::new().unwrap();
        let kind_dir = tmp.path().join("memory_op");
        tokio::fs::create_dir_all(&kind_dir).await.unwrap();
        // Broken indentation under `proposal:` keeps it from parsing as a
        // mapping; serde_yaml surfaces the error.
        write(
            &kind_dir,
            "broken.yaml",
            "description: bad\nproposal:\n  target: x\n   reasoning: y\n",
        )
        .await;
        let err = load_eval_set(tmp.path(), EvolutionKind::MemoryOp)
            .await
            .unwrap_err();
        assert!(
            matches!(err, EvalLoadError::ParseFailure { .. }),
            "got {err:?}"
        );
    }

    #[tokio::test]
    async fn load_eval_set_rejects_kind_mismatch() {
        let tmp = TempDir::new().unwrap();
        let kind_dir = tmp.path().join("memory_op");
        tokio::fs::create_dir_all(&kind_dir).await.unwrap();
        write(
            &kind_dir,
            "wrong.yaml",
            r#"
kind: skill_update
description: wrong kind
proposal:
  target: irrelevant
  reasoning: irrelevant
expected:
  outcome: no_op
"#,
        )
        .await;
        let err = load_eval_set(tmp.path(), EvolutionKind::MemoryOp)
            .await
            .unwrap_err();
        assert!(
            matches!(
                err,
                EvalLoadError::KindMismatch { ref found, .. } if found == "skill_update"
            ),
            "got {err:?}"
        );
    }

    #[tokio::test]
    async fn load_eval_set_rejects_empty_dir() {
        let tmp = TempDir::new().unwrap();
        tokio::fs::create_dir_all(tmp.path().join("memory_op"))
            .await
            .unwrap();
        let err = load_eval_set(tmp.path(), EvolutionKind::MemoryOp)
            .await
            .unwrap_err();
        assert!(matches!(err, EvalLoadError::EmptySet { .. }), "got {err:?}");
    }

    #[tokio::test]
    async fn load_eval_set_skips_underscore_prefixed() {
        let tmp = TempDir::new().unwrap();
        let kind_dir = tmp.path().join("memory_op");
        tokio::fs::create_dir_all(&kind_dir).await.unwrap();
        let valid = r#"
description: real case
proposal:
  target: merge_chunks:1,2
  reasoning: dupes
expected:
  outcome: no_op
"#;
        write(&kind_dir, "real.yaml", valid).await;
        // `_draft.yaml` is intentionally bogus; loader must skip it.
        write(&kind_dir, "_draft.yaml", "this is not yaml :::").await;
        let set = load_eval_set(tmp.path(), EvolutionKind::MemoryOp)
            .await
            .unwrap();
        assert_eq!(set.cases.len(), 1);
        assert_eq!(set.cases[0].name, "real");
    }

    #[tokio::test]
    async fn load_eval_set_sorts_cases_by_name() {
        let tmp = TempDir::new().unwrap();
        let kind_dir = tmp.path().join("memory_op");
        tokio::fs::create_dir_all(&kind_dir).await.unwrap();
        let body = |name: &str| {
            format!(
                r#"
name: {name}
description: ordering check
proposal:
  target: t
  reasoning: r
expected:
  outcome: no_op
"#
            )
        };
        // Write out of order on disk; loader must sort by `name`.
        write(&kind_dir, "z.yaml", &body("zebra")).await;
        write(&kind_dir, "a.yaml", &body("alpha")).await;
        write(&kind_dir, "m.yaml", &body("mango")).await;
        let set = load_eval_set(tmp.path(), EvolutionKind::MemoryOp)
            .await
            .unwrap();
        let names: Vec<&str> = set.cases.iter().map(|c| c.name.as_str()).collect();
        assert_eq!(names, vec!["alpha", "mango", "zebra"]);
    }
}

File: python/packages/corlinman-shadow-tester/src/corlinman_shadow_tester/eval.py
"""Eval-case spec + YAML loader for ShadowTester.

Ported 1:1 from ``rust/crates/corlinman-shadow-tester/src/eval.rs``.

Cases live as YAML files under ``<eval_set_dir>/<kind>/*.yaml``. The
per-kind subdir is the contract: it lets :func:`load_eval_set` default
``kind`` from the path so authors don't have to repeat themselves, and
makes ``ls memory_op/`` the way an operator audits coverage.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml
from corlinman_evolution_store import EvolutionKind, EvolutionRisk


DEFAULT_LATENCY_MS_MAX = 500


@dataclass
class ProposalSpec:
    """Minimum proposal data the simulator needs.

    The runner assembles a full :class:`EvolutionProposal` by attaching a
    generated id + timestamps when it dispatches to a simulator.
    """

    target: str
    reasoning: str
    # Shadow only fires for medium/high; default High keeps fixtures
    # in-scope by default.
    risk: EvolutionRisk = EvolutionRisk.HIGH
    signal_ids: list[int] = field(default_factory=list)
    # Unified-diff payload — ``skill_update`` ships an ``__APPEND__`` hunk
    # here. memory_op / tag_rebalance leave it empty.
    diff: str = ""


@dataclass
class ExpectedOutcome:
    """What the simulator should observe after replaying the proposal.

    ``outcome`` is the discriminator (snake_case string). The other
    fields are populated per-variant; absent fields default to ``None``
    (the variant inspection layer is the simulator).

    Variants (matching the Rust ``ExpectedOutcome`` enum tag values):

    - ``merged`` — memory_op merge that consumed ``rows_merged`` chunks
      and kept ``surviving_chunk_id`` as the canonical row.
    - ``no_op`` — memory_op detected a bogus / unsafe target.
    - ``tag_merged`` — tag_rebalance executed; ``src_path`` gone and
      ``parent_id`` now owns its ``chunk_tags`` rows.
    - ``tag_no_op`` — tag_rebalance target path didn't resolve.
    - ``skill_updated`` — skill_update file appended; final body must
      contain ``content_includes`` as a substring.
    - ``skill_no_op`` — skill_update rejected (unsupported diff shape,
      missing file, invalid target).
    """

    outcome: str
    rows_merged: int | None = None
    surviving_chunk_id: int | None = None
    src_path: str | None = None
    parent_id: int | None = None
    moved_chunk_count: int | None = None
    file: str | None = None
    content_includes: str | None = None
    latency_ms_max: int = DEFAULT_LATENCY_MS_MAX

    _ALLOWED: frozenset[str] = frozenset(
        {
            "merged",
            "no_op",
            "tag_merged",
            "tag_no_op",
            "skill_updated",
            "skill_no_op",
        }
    )

    def __post_init__(self) -> None:
        if self.outcome not in self._ALLOWED:
            raise EvalParseError(f"unknown expected outcome: {self.outcome!r}")


@dataclass
class EvalCase:
    """One YAML-defined test case for one kind.

    ``kb_seed`` runs raw SQL against a tempdir copy of ``kb.sqlite``
    before the proposal is shadowed. Normally ``INSERT INTO chunks ...``;
    cases that want a from-scratch fixture can include ``CREATE TABLE``
    first.
    """

    description: str
    proposal: ProposalSpec
    expected: ExpectedOutcome
    # Defaults to the YAML file stem in :func:`load_eval_set`.
    name: str = ""
    # Defaults to the directory's kind in :func:`load_eval_set`.
    kind: EvolutionKind | None = None
    kb_seed: list[str] = field(default_factory=list)
    # ``<basename> -> file body`` map written into a runner-managed
    # per-case ``<tempdir>/skills/`` directory before the simulator runs.
    # Empty for kinds that don't touch ``skills/``.
    skill_seed: dict[str, str] = field(default_factory=dict)


@dataclass
class EvalSet:
    """All cases loaded from one ``<dir>/<kind>/`` subdir."""

    kind: EvolutionKind
    cases: list[EvalCase]
    loaded_from: Path


@dataclass
class EvalRunResult:
    """Per-case run result; metrics shape is simulator-defined."""

    case_name: str
    passed: bool
    metrics: dict[str, Any] = field(default_factory=dict)
    error: str | None = None


# ---------------------------------------------------------------------------
# Errors. Empty-set is intentionally an error: a misconfigured path that
# silently shadows zero cases would look "green" forever.
# ---------------------------------------------------------------------------


class EvalLoadError(RuntimeError):
    """Base class for every loader error."""


class MissingDirError(EvalLoadError):
    def __init__(self, path: Path) -> None:
        super().__init__(f"eval-set dir missing: {path}")
        self.path = path


class IoError(EvalLoadError):
    def __init__(self, path: Path, source: Exception) -> None:
        super().__init__(f"io error reading {path}: {source}")
        self.path = path
        self.source = source


class ParseFailure(EvalLoadError):
    def __init__(self, file: Path, reason: str) -> None:
        super().__init__(f"parse failure in {file}: {reason}")
        self.file = file
        self.reason = reason


class KindMismatch(EvalLoadError):
    def __init__(self, file: Path, expected: EvolutionKind, found: str) -> None:
        super().__init__(
            f"kind mismatch in {file}: expected {expected!r}, found '{found}'"
        )
        self.file = file
        self.expected = expected
        self.found = found


class EmptySet(EvalLoadError):
    def __init__(self, dir_: Path, kind: EvolutionKind) -> None:
        super().__init__(f"no eval cases found for kind {kind!r} under {dir_}")
        self.dir = dir_
        self.kind = kind


class EvalParseError(EvalLoadError):
    """Raised when a YAML doc cannot be coerced into an :class:`EvalCase`."""


# ---------------------------------------------------------------------------
# Loader
# ---------------------------------------------------------------------------


def _parse_case(raw: Any) -> EvalCase:
    """Coerce a parsed YAML document into :class:`EvalCase`.

    Mirrors the Rust ``serde(default)`` shape — missing fields default to
    sensible empties. Throws :class:`EvalParseError` on shape mismatch.
    """
    if not isinstance(raw, dict):
        raise EvalParseError(f"top-level YAML must be a mapping, got {type(raw).__name__}")

    description = raw.get("description")
    if not isinstance(description, str):
        raise EvalParseError("missing required str 'description'")

    proposal_raw = raw.get("proposal")
    if not isinstance(proposal_raw, dict):
        raise EvalParseError("missing required mapping 'proposal'")
    target = proposal_raw.get("target")
    reasoning = proposal_raw.get("reasoning")
    if not isinstance(target, str):
        raise EvalParseError("proposal.target must be a string")
    if not isinstance(reasoning, str):
        raise EvalParseError("proposal.reasoning must be a string")
    risk_raw = proposal_raw.get("risk", "high")
    risk = EvolutionRisk.from_str(str(risk_raw))
    signal_ids_raw = proposal_raw.get("signal_ids", [])
    if not isinstance(signal_ids_raw, list):
        raise EvalParseError("proposal.signal_ids must be a list")
    signal_ids = [int(x) for x in signal_ids_raw]
    diff = proposal_raw.get("diff", "")
    if not isinstance(diff, str):
        raise EvalParseError("proposal.diff must be a string")

    expected_raw = raw.get("expected")
    if not isinstance(expected_raw, dict):
        raise EvalParseError("missing required mapping 'expected'")
    outcome = expected_raw.get("outcome")
    if not isinstance(outcome, str):
        raise EvalParseError("expected.outcome must be a string")
    expected = ExpectedOutcome(
        outcome=outcome,
        rows_merged=expected_raw.get("rows_merged"),
        surviving_chunk_id=expected_raw.get("surviving_chunk_id"),
        src_path=expected_raw.get("src_path"),
        parent_id=expected_raw.get("parent_id"),
        moved_chunk_count=expected_raw.get("moved_chunk_count"),
        file=expected_raw.get("file"),
        content_includes=expected_raw.get("content_includes"),
        latency_ms_max=int(expected_raw.get("latency_ms_max", DEFAULT_LATENCY_MS_MAX)),
    )

    name = raw.get("name", "") or ""
    kind_raw = raw.get("kind")
    kind: EvolutionKind | None = None
    if kind_raw is not None:
        kind = EvolutionKind.from_str(str(kind_raw))

    kb_seed_raw = raw.get("kb_seed", [])
    if not isinstance(kb_seed_raw, list):
        raise EvalParseError("kb_seed must be a list of strings")
    kb_seed = [str(x) for x in kb_seed_raw]

    skill_seed_raw = raw.get("skill_seed", {}) or {}
    if not isinstance(skill_seed_raw, dict):
        raise EvalParseError("skill_seed must be a mapping")
    skill_seed = {str(k): str(v) for k, v in skill_seed_raw.items()}

    return EvalCase(
        description=description,
        proposal=ProposalSpec(
            target=target,
            reasoning=reasoning,
            risk=risk,
            signal_ids=signal_ids,
            diff=diff,
        ),
        expected=expected,
        name=str(name),
        kind=kind,
        kb_seed=kb_seed,
        skill_seed=skill_seed,
    )


async def load_eval_set(eval_set_dir: Path, kind: EvolutionKind) -> EvalSet:
    """Load every ``*.yaml``/``*.yml`` file from ``<eval_set_dir>/<kind>/``.

    Files prefixed with ``_`` are skipped (drafts). Non-recursive. Cases
    are sorted by ``name`` for deterministic ordering across runs.

    The Rust implementation is fully async via ``tokio::fs``; we delegate
    filesystem and YAML parsing to a thread (via ``asyncio.to_thread``)
    so the coroutine yields and matches the async surface.
    """
    return await asyncio.to_thread(_load_eval_set_sync, eval_set_dir, kind)


def _load_eval_set_sync(eval_set_dir: Path, kind: EvolutionKind) -> EvalSet:
    dir_ = eval_set_dir / kind.as_str()
    try:
        exists = dir_.exists()
    except OSError as exc:
        raise IoError(dir_, exc) from exc
    if not exists:
        raise MissingDirError(dir_)

    yaml_files: list[Path] = []
    try:
        for entry in dir_.iterdir():
            name = entry.name
            if name.startswith("_"):
                continue
            suffix = entry.suffix.lower()
            if suffix not in (".yaml", ".yml"):
                continue
            if not entry.is_file():
                continue
            yaml_files.append(entry)
    except OSError as exc:
        raise IoError(dir_, exc) from exc

    cases: list[EvalCase] = []
    for file in yaml_files:
        try:
            text = file.read_text(encoding="utf-8")
        except OSError as exc:
            raise IoError(file, exc) from exc
        try:
            raw = yaml.safe_load(text)
        except yaml.YAMLError as exc:
            raise ParseFailure(file, str(exc)) from exc
        try:
            case = _parse_case(raw)
        except EvalParseError as exc:
            raise ParseFailure(file, str(exc)) from exc

        # Reject explicit-but-wrong kinds; default the unset case to the dir's kind.
        if case.kind is not None and case.kind != kind:
            raise KindMismatch(file, kind, case.kind.as_str())
        case.kind = kind

        if not case.name:
            case.name = file.stem or "unnamed"
        cases.append(case)

    if not cases:
        raise EmptySet(dir_, kind)

    cases.sort(key=lambda c: c.name)
    return EvalSet(kind=kind, cases=cases, loaded_from=dir_)


__all__ = [
    "DEFAULT_LATENCY_MS_MAX",
    "EmptySet",
    "EvalCase",
    "EvalLoadError",
    "EvalParseError",
    "EvalRunResult",
    "EvalSet",
    "ExpectedOutcome",
    "IoError",
    "KindMismatch",
    "MissingDirError",
    "ParseFailure",
    "ProposalSpec",
    "load_eval_set",
]
